### ISSUE TITLE
sql: TransactionStatusError: transaction deadline exceeded

### ISSUE THREAD

**Issue by jordanlewis** [cockroachdb/cockroach#18684]
When running TPCC with more than a single thread, occasionally a transaction will fail with the following error:

` pq: TransactionStatusError: transaction deadline exceeded`

What is this about? What is the referenced deadline? It seems like a bug that this message is propagated to the user, but what internal condition is causing this scenario and why?

Any ideas? cc @andreimatei @bdarnell

**Comment by bdarnell** [comment]
I believe that message means a table descriptor lease has expired. cc @vivekmenezes @lego

**Comment by lego** [comment]
@jordanlewis how frequent is the error? If it is frequent enough it will hopefully be easy to test whether the lease expiring and re-acquisition is the issue since I have an in-flight PR right now (relevant PR #18606, issue #17227).

That is peculiar that it happens specifically with more than one thread. There are a few places in the lease acquisition where I can see race conditions causing starvation but only if the lease is immediately released, or _maybe_ under heavy load. This could go either two ways: the above PR fixes it because lease acquisition is taking too long, or it's a race-condition causing a thread to be starved. I would think the former would happen with also one thread though, so it's less likely to be that.

**Comment by vivekmenezes** [comment]
This bug is happening because we are using a lease very close to its expiration deadline (we changed that recently to do that in https://github.com/cockroachdb/cockroach/pull/16842 by taking out the notion of MinLeaseDuration). I suspect we are seeing this here because a transaction with SNAPSHOT ISOLATION is using a table descriptor with an unexpired lease but is having its COMMIT pushed beyond the lease expiration (which is what the txn expiration is set to).

lego's PR will fix this because it will reacquire a lease way before it expires and thus transactions will use leases with expirations further out into the future (like a minute).

**Comment by jordanlewis** [comment]
All transactions in TPCC are set to Serializable isolation.

**Comment by vivekmenezes** [comment]
if you can log the txn.OrigTimestamp() and expiration in getTableVersion() after the call to AcquireByName() and panic instead of returning the error with the deadline timestamp and the txn timestamp we might get a better clue. Thanks!

**Comment by vivekmenezes** [comment]
of course there is also the possibility that we should be using OrigTimestamp() when acquiring leases. I think I need to ask around as to what API to use to get the COMMIT timestamp for a transaction as it's running

**Comment by vivekmenezes** [comment]
Turns out using a timestamp near the deadline, while valid when picked, is bad in general, because the transaction timestamp can be pushed in the background by another transaction and a COMMIT will see the new pushed timestamp (when it reads the transaction record from the store) and return a violation of the deadline.

So  #18606 is a good fix for this issue

**Comment by jordanlewis** [comment]
@lego it's pretty easy to reproduce that error with TPCC. Perhaps you could try it out on your branch? Instructions:

1. Start a local cockroach instance
2. Clone the `cockroachdb/loadgen` repo, enter the `tpcc` directory, `go build`
3. Load the data with `./tpcc -load`
4. Once the data's loaded, it'll start the load generator automatically. Ctrl-C
5. Run with high concurrency: `./tpcc -concurrency=128`. That usually triggers the above error for me.

**Comment by andreimatei** [comment]
#18606 seems like an improvement to things but ideally we wouldn't rely on some async process to ensure that our transactions can commit.
The general idea, I believe, should be that you can verify if a timestamp is valid or not for a txn (wrt schema changes) whenever you want. So, if some layer is about to say "deadline exceeded", it should also give SQL a chance to say "no, actually, the txn timestamp is still fine because I've verified that the descriptors are used are valid for the timestamp in question". I don't know what the best way to transcribe that into code would be, but we can investigate.

**Comment by bdarnell** [comment]
Prior to #16842, we checked the time left on the lease and would (synchronously) renew it if there was less than a minute left. Why was this removed? Without it it's easy to run into this error because there's no guarantee that there's a useful amount of time left on the lease. This is a regression compared to 1.0 so I'm tagging this for 1.1. 

I think we need both MinLeaseDuration and #18606: synchronously renew the lease when it's too close to expiring, and asynchronously renew it somewhat before it reaches that point. 

Additionally, to support arbitrarily-long-running transactions, we would need some way for the renewed lease to be fed into the coordinator (and from there probably written to the transaction record via HeartbeatTxn). That's much lower priority, though.

**Comment by andreimatei** [comment]
1. First of all, let's agree that these txn deadlines should only generate errors for SNAPSHOT transactions. For SERIALIZABLE txns, the `TransactionRetryError` caused by the push should take precedence. It is, in my opinion, a bug that this line generating the deadline error:
https://github.com/cockroachdb/cockroach/blob/57d142d15ad45359eac7521448fb7cbc9cbc487d/pkg/storage/replica_command.go#L769
is above the lines just below that would generate the `TransactionRetryError`. I think this is the first thing that we should fix; then all the discussion remaining would be restricted to SNAPSHOT txns.

2. Second of all, I think we should turn these deadline errors into retriable errors so that SQL retries.


Third, I'd like to  respond to Ben's message:


> Without it it's easy to run into this error because there's no guarantee that there's a useful amount of time left on the lease. This is a regression compared to 1.0 so I'm tagging this for 1.1.

What you say is true, but just in case we're not all on the same page, note that this deadline does not prevent "arbitrarily-long-running transactions". It just prevents arbitrarily-big-pushes. You can have a transaction run arbitrarily long without changing its timestamp (and the deadline is checked against the txn timestamp, not the clock).

So I don't think that what we want to do is synchronously renew leases (or, rather, I'd ask you when do you renew the lease - when it's close to expiring or when the txn in question has been pushed close to the deadline, or even beyond it?). In all this discussion, keep in mind that the client may not even know that the transaction has been pushed.

Another thing I want to note is that, for a transaction to be able to commit at a particular timestamp, it doesn't necessarily need for the gateway to have valid leases on the descriptors that it used. Remember from the newer version of the descriptors and leases RFC that the conditions sufficient for a transaction to commit are divorced from the leases; they're about the validity window of the descriptors it used.

How about this: we augment the `client.Txn` interface with a callback through which KV can ask whomever has set the txn deadline (i.e. SQL) about whether the txn can commit at a timestamp *T_pushed* beyond the originally-set deadline. SQL would set this callback to check whether *T_pushed* is within the validity window of all the descriptors; it would do this by looking at the descriptor versions that the node knows about and, where it doesn't have enough information, it'd read the database. Reading the database may, as a byproduct, also renew a lease, but that's not essential.
The callback would be called by `txn.Send()` whenever it gets a "deadline exceeded" error. We'd rely on these errors only being generated by `EndTransaction` and so, if the callback says that we're still golden at the new timestamp, then `client.Txn` would re-issue the `EndTransaction` (just this, no other requests from the last batch) with an updated deadline.

Otherwise, what you're hinting at around renewing leases in the background and updating txn deadlines in the background sounds good - it'd prevent the "deadline error" from happening in most cases. But note that a lease expiration is not trivially translatable into a txn deadline: the deadline for a txn is set based on a set of descriptors not just one (and also, again, based on validity windows which may not be related to leases).

**Comment by bdarnell** [comment]
> First of all, let's agree that these txn deadlines should only generate errors for SNAPSHOT transactions.

Agreed.

> So I don't think that what we want to do is synchronously renew leases (or, rather, I'd ask you when do you renew the lease - when it's close to expiring or when the txn in question has been pushed close to the deadline, or even beyond it?)

I'd only synchronously get the lease when it is first accessed. After that we can asynchronously renew before expiration for as long as the transaction's open. 

> The callback would be called by txn.Send() whenever it gets a "deadline exceeded" error. We'd rely on these errors only being generated by EndTransaction and so, if the callback says that we're still golden at the new timestamp, then client.Txn would re-issue the EndTransaction (just this, no other requests from the last batch) with an updated deadline.

I don't like the idea of retrying after an EndTransaction. Per comments in the code, we'd like to formally abort the transaction when this happens (but this is currently prevented by other issues). If we did that, we wouldn't be able to retry. Maybe we just change the comments and guarantee that attempting to commit with a failed deadline is safe, but this seems subtle to me.

**Comment by andreimatei** [comment]
> I don't like the idea of retrying after an EndTransaction. Per comments in the code, we'd like to formally abort the transaction when this happens (but this is currently prevented by other issues). If we did that, we wouldn't be able to retry. Maybe we just change the comments and guarantee that attempting to commit with a failed deadline is safe, but this seems subtle to me.

The comments you're referring to were written (presumably) under the belief that what we want to do when a transaction's deadline was exceeded is to abort the transaction. I don't think they should be used to inform decisions in the situation where we decide that we, in fact, do not want to abort.
To take this further, I in fact believe that we *never* want to abort; in the worst case, the transaction should be prepared for a retry (increment epoch, leave intents in place).

**Comment by lego** [comment]
Merged https://github.com/cockroachdb/cockroach/pull/18824 which will help resolve the immediate issue at hand, the `TransactionStatusError` error. Do we want to close this issue and open another regarding the other problem around serializable restarts?

**Comment by andreimatei** [comment]
This issue has not been resolved by #18824; your PR has only reduced the incidence of the problem (a txn pushed sufficiently will still encounter the same error; you've just made it so that the required push is now expected to be larger than before).
For Serializable txns, I've sent a PR that makes them never encounter this error; it's linked above.


We can remove the 1.1 milestone, but otherwise the issue is still valid.
But, it seems to me like the patch in question still needs some work. I've left some comments on the original PR. If I'm right, I think we should rollback the cherry-pick until the patch gets a bit more love.

**Comment by vivekmenezes** [comment]
This issue has been fixed for the SERIALIZABLE transaction case. Keeping this open for the SNAPSHOT isolation case which is less important.

**Comment by lego** [comment]
Seems the problem for SERIALIZABLE transactions has not gone away yet and it's still happening around when the expires after https://github.com/cockroachdb/cockroach/pull/18804.

**Comment by andreimatei** [comment]
Yup, badness reproduces easily with TPCC (particularly if I make the lease duration lower).

Check out a txn running an `EndTransaction` and its deadline.
The transaction has ts=1506632508.129725668,1 orig=**1506632508**.129725668,1 .
The `EndTransaction` has deadline **1506632504**.748866348,0.
That's obviously no bueno - the deadline is below the `OrigTimestamp`. From the log dump below we can see that the deadline is around what appears to be the `OrigTimestamp` of a previous attempt.

I suspect the problem has to do with the interaction between transaction restarts and the deadline/the set of descriptors used. On a restart, the deadline is supposed to be reset here:
https://github.com/cockroachdb/cockroach/blob/bcdd4ca16c57fbda973287494b54c32ae194ed2b/pkg/sql/table.go#L270
Perhaps that doesn't happen for some reason.



```
I170928 21:01:48.301009 179379 storage/replica_command.go:786  [n1,s1,r19/1:/Table/5{2-3}] !!! deadline exceeded. txn:
"sql txn" id=2948385d key=/Table/52/1/0/6/0 rw=true pri=0.03825747 iso=SERIALIZABLE stat=PENDING epo=1 ts=1506632508.129725668,1 orig=**1506632508**.129725668,1 max=1506632504.335166591,0 wto=false rop=false seq=134 dead=%!s(MISSING)
reply.Txn:
"sql txn" id=2948385d key=/Table/52/1/0/6/0 rw=true pri=0.03825747 iso=SERIALIZABLE stat=PENDING epo=1 ts=1506632508.129725668,1 orig=1506632504.335166591,0 max=1506632504.335166591,0 wto=false rop=false seq=38 dead=%!s(MISSING)
deadline: **1506632504**.748866348,0
```

**Comment by vivekmenezes** [comment]
So it looks like the SQL layer is setting the deadline (txn.deadline) for the txn correctly, but when the SQL layer runs a commit on the transaction it sends another deadline (through txn.deadline). This is looking like a really bad bug.

**Comment by andreimatei** [comment]
I don't quite have the full story yet, but the problem is about the a transaction that gets restarted (either through a TransactionRestartError or a TransactionRetryError - i.e. either with the same id or with a new one), and keeping the old deadline instead of having its deadline be reset the first time a table is accessed. Somehow, after the restart, the session's `TableCollection.timestamp` is not initialized, but the txn is initialized (it was initialized before the restart and stays the same). Exactly how that happens, I don't know yet - the deadline and `tc.timestamp` are supposed to be updated together, although the code is a bit convoluted.
In any case, won't be much longer...

**Comment by andreimatei** [comment]
OK, I've got it. The problem is this call to `TableCollection.releaseTables()` in `execSchemaChanges()`:

https://github.com/cockroachdb/cockroach/blob/2bac2d407e21dd6b71c0b93ccb50356900f3a612/pkg/sql/session.go#L1318

This is the point where `tc.timestamp` gets reset, but `txn.Deadline` doesn't. Because `tc.Timestamp` is reset, the deadline will never get reset, as the condition below in `tc.resetForTxnRetry()` will never fire:

https://github.com/cockroachdb/cockroach/blob/2bac2d407e21dd6b71c0b93ccb50356900f3a612/pkg/sql/table.go#L270

I'll do something about it.

**Comment by jordanlewis** [comment]
@andreimatei I don't understand exactly - are you saying that the problem is isolated to `execSchemaChanges()`? In that case, why does this problem appear in the schema-change-less `tpcc` workload?

**Comment by vivekmenezes** [comment]
```execSchemaChanges(``` is always called. It just never does anything normally. But has that lovely side effect that andrei mentioned.

**Comment by andreimatei** [comment]
To summarize the current state of affairs:
- we've fixed an issue causing the KV to return the "transaction deadline exceeded" error instead of a restart error for serializable transactions pushed beyond their deadline
- we've fixed an issue causing restarted transactions to not have their deadline updated - in turn causing them to become increasingly likely to exceed their deadline as they were getting pushed and restarting
- #18844 is in review - it's supposed to have leases "refreshed" such that, on average, their expected to have more time left whenever a txn uses them -> so snapshot transactions are, on average, going to allow larger pushes before they exceed their deadline

What's left for this bug is better handling for the situation where a snapshot txn is pushed beyond its deadline in spite of the various improvements listed above.
1. If there's been no schema changes, we should find a way extend the deadline beyond what it was originally set to / allow the transaction commit.
2. If a transaction can't commit because of the deadline, it should be a retriable error.
3. Deadlines should not be enforced for read-only transactions.

**Comment by vivekmenezes** [comment]
This issue should be closed after we've deprecated SNAPSHOT ISOLATION

**Comment by jordanlewis** [comment]
Wait, we still get these all the time in TPCC, and no one is doing schema changes or SNAPSHOT isolation transactions during it.

**Comment by jordanlewis** [comment]
In case it helps, here's a stack trace of how the error is created:

```
github.com/cockroachdb/cockroach/pkg/storage.evalEndTransaction(0x3550940, 0xc42ef759b0, 0x7fb0395fb818, 0xc465499200, 0x3564b40, 0xc420667200, 0x1514de0ad4697e31, 0x1, 0x100000001, 0x1, ...)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_command.go:362 +0x1d2f
github.com/cockroachdb/cockroach/pkg/storage.evaluateCommand(0x3550940, 0xc42ef759b0, 0xc4363665a0, 0x8, 0x0, 0x7fb0395fb818, 0xc465499200, 0x3564b40, 0xc420667200, 0xc43347ed90, ...)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_command.go:101 +0x30c
github.com/cockroachdb/cockroach/pkg/storage.evaluateBatch(0x3550940, 0xc42ef759b0, 0xc4363665a0, 0x8, 0x7fb0395fb818, 0xc465499200, 0x3564b40, 0xc420667200, 0xc43347ed90, 0x1514de0ad4697e31, ...)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:5463 +0x3ad
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).evaluateWriteBatchWithLocalRetries(0xc420667200, 0x3550940, 0xc42ef759b0, 0xc4363665a0, 0x8, 0x3564b40, 0xc420667200, 0xc43347ed90, 0x1514de0ad4697e31, 0x1, ...)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:5250 +0x1da
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).evaluateWriteBatch(0xc420667200, 0x3550940, 0xc42ef759b0, 0xc4363665a0, 0x8, 0x1514de0ad4697e31, 0x1, 0x100000001, 0x1, 0x16, ...)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:5226 +0x20f
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).evaluateProposalInner(0xc420667200, 0x3550940, 0xc42ef759b0, 0xc4363665a0, 0x8, 0x1514de0ad4697e31, 0x1, 0x100000001, 0x1, 0x16, ...)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:5065 +0x16a
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).evaluateProposal(0xc420667200, 0x3550940, 0xc42ef759b0, 0xc4363665a0, 0x8, 0x1514de0ad4697e31, 0x1, 0x100000001, 0x1, 0x16, ...)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2856 +0xe0
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).requestToProposal(0xc420667200, 0x3550940, 0xc42ef759b0, 0xc4363665a0, 0x8, 0x1514de0ad4697e31, 0x1, 0x100000001, 0x1, 0x16, ...)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2813 +0x19f
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).propose(0xc420667200, 0x3550940, 0xc42ef759b0, 0x1514dc86a8abba1a, 0x0, 0x0, 0x100000001, 0x1, 0x0, 0xc42b44f110, ...)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:3008 +0x4cf
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).tryExecuteWriteBatch(0xc420667200, 0x3550940, 0xc42ef759b0, 0x1514de0ad4697e31, 0x1, 0x100000001, 0x1, 0x16, 0x3ff0000000000000, 0xc4555afe00, ...)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2710 +0x3eb
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).executeWriteBatch(0xc420667200, 0x3550940, 0xc42ef759b0, 0x1514de0ad4697e31, 0x1, 0x100000001, 0x1, 0x16, 0x3ff0000000000000, 0xc4555afe00, ...)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2560 +0xae
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).Send(0xc420667200, 0x3550940, 0xc42ef759b0, 0x1514de0ad4697e31, 0x1, 0x100000001, 0x1, 0x16, 0x3ff0000000000000, 0xc4555afe00, ...)
```

It's pretty straightforward for me to reproduce this, so if anyone has an idea for what other kinds of information to collect, please let me know.

**Comment by vivekmenezes** [comment]
@jordanlewis are you able to readily reproduce this? I'm unable to reproduce this and the only way I can imagine this can happen is under the following circumstances

1. A query runs  and acquires a table lease
2. No other queries run so the lease is not reacquired asynchronously
3. Another query runs very close to the expiration time of the lease
4. this query is retried internally by the kv layer pushing the commit timestamp to after the OrigTimestamp.

while this might be happening I seriously doubt the above is readily reproducible and is the situation you are seeing.

**Comment by vivekmenezes** [comment]
ah ha I see you used -concurrency=128. let me give that a try

**Comment by vivekmenezes** [comment]
I'm unable to repro it after running tpcc for 13 minutes with ```-concurrency=128```

**Comment by jordanlewis** [comment]
Yes, this is easily reproducible with the TPCC-1000 dataset.  How were you trying to reproduce it?

**Comment by jordanlewis** [comment]
Can we mark this error as retryable? If that's possible then we can close this.

**Comment by andreimatei** [comment]
From the updates above it sounds like we would have expected deadlines to be turned into retriable errors in some cases. I think this deserves investigation; I'll see what's going on.

**Comment by andreimatei** [comment]
I completely missed the point of Vivek's update above; the point was that there might have been a regression for Serializable txns ever since the introduction of the `RefreshRequest` - whereas it used to be the case that pushed serializable txns get a retriable error (this was previously fixed in https://github.com/andreimatei/cockroach/commit/9abaa0dce684825f6cef41359ae082607dfddc9e), now it may be the case that pushed transactions no longer get the retriable error (if their `Refresh Request` succeeds) but can still get the deadline exceeded error.

Besides the possible future work listed in https://github.com/cockroachdb/cockroach/issues/18684#issuecomment-333647564 ,  I think there are a couple of options for resolving the regression:

1. Make the refresh code aware of the deadline and refuse to ever push the timestamp beyond this. The problem here is that currently the deadline is not an attribute of the txn, but of the `EndTransactionRequest`, and so generally the refresh code can't see it, but maybe we can change this.
2. Live with the regression, but make the deadline exceeded a retriable error. I don't know if it'd be a good idea to that at the storage level, or if we should convert the error at the SQL level (given that this deadline thing is only retriable when then client is SQL... but on the other hand only SQL uses the feature...)

@spencerkimball , do you have any thoughts, particularly on 1) ?

**Comment by spencerkimball** [comment]
On first glance, I see no reason why the deadline exceeded error shouldn't be retryable – this was something I was wondering about previously.

I'll send a PR.

**Comment by andreimatei** [comment]
Well, the question is, who's the right guy to say that it's retriable? SQL is for its use of the deadline, since on a retry it'll acquire new descriptors with new deadlines. Is storage also able to say that? If we were to introduce other uses of the deadline...

**Comment by spencerkimball** [comment]
Well my plan is to return a new reason with `TransactionRetryError`: `ERROR_DEADLINE_EXCEEDED`, which can be used by higher levels to make the call.

**Comment by vivekmenezes** [comment]
Sounds good. I remember the last time I tried making this a retryable error many months ago, it didn't clean up intents so it would hit the old intents on retry. I hope you can also cleanup the intents. Thanks!

**Comment by andreimatei** [comment]
> Well my plan is to return a new reason with TransactionRetryError: ERROR_DEADLINE_EXCEEDED, which can be used by higher levels to make the call.

Except that, for `TransactionRetryError`, higher layers will always retry - the `TxnCS` will turn it into a `HandledRetryableTxnError` and `db.Txn()`, for example, will retry.
I think SQL should recognize an otherwise non-retriable error (i.e. a pErr without the `TransactionRestart` field set).

Stepping out again, I'm not sure I can convince myself that the deadline exceeded error needs to exist at all, at least for serializable txns. As I was thinking above, it is the refresh that bumps the timestamp and so it could make some effort to not bump it past this deadline. If the refresh wouldn't
bump us, then higher layers would get another (retriable) error - the one causing the push or the read uncertainty. Which would be more honest. Except then again, it could be useful in the future for SQL to particularly distinguish deadline exceeded errors (where the refresh succeeded), because in the vast majority of cases SQL doesn't actually have to retry - it just needs to check if the descriptors it used are still valid. But then ideally SQL would have all the information - the original error causing a push and the information that a refresh would succeed if it wasn't for the deadline.
I dunno.

**Comment by vivekmenezes** [comment]
@andreimatei I think this error is so rare (I've not been able to reproduce it) it's not worth attempting to optimize how we handle it. I'm in favor of running it through a retry at least at this stage.

**Comment by spencerkimball** [comment]
It’s apparently not rare in the TPCC workload. @jordanlewis?

**Comment by jordanlewis** [comment]
Yeah, it's not rare. @vivekmenezes if you can't reproduce it, there must be something different from your test setup than an actual cluster with a lot of data and ranges etc. Here's a run from earlier  - got the error in 25 seconds.

```
jordan@jordan-tpcc-0025:~$ ./tpcc --warehouses=3000 --duration=5m --split --scatter  $nodeaddrs
_time______opName__ops/s(inst)__ops/s(cum)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
   2s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
   3s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
   4s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
   5s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
   6s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
   7s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
   8s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
   9s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
  10s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
  11s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
  12s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
  13s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
  14s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
  15s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
  16s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
  17s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
  18s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
  19s    newOrder          0.0         0.0      0.0      0.0      0.0      0.0      0.0
  20s    newOrder        349.0        17.8    268.4    335.5    352.3    419.4    469.8
  21s    newOrder        371.2        34.8    285.2    335.5    369.1    402.7    520.1
_time______opName__ops/s(inst)__ops/s(cum)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  22s    newOrder        263.9        45.4    318.8    453.0    503.3    570.4    604.0
  23s    newOrder        387.9        60.5    419.4    637.5    671.1    704.6    838.9
  24s    newOrder        463.8        77.5    302.0    369.1    385.9    738.2   1677.7
  25s    newOrder        400.2        90.7    318.8    436.2    436.2    486.5    671.1
2018/03/08 19:27:23 error in newOrder: pq: TransactionStatusError: transaction deadline exceeded

**Comment by bdarnell** [comment]
Prior to #16842, we had a `MinLeaseDuration` in `sql/lease.go` that ensured that if there was less than a minute remaining on the table lease, we'd wait for the lease to update instead of running with a short deadline. This appears to have been lost and we now have transactions allowed to run with less than 25s left on their lease. It looks like we may need to bring back some form of MinLeaseDuration.

**Comment by andreimatei** [comment]
I don't think I like a `MinLeaseDuration`; txns that aren't pushed don't need any particular amount of time left on the lease... And even for pushed txns, if we took the effort to "recalculate" the deadline, they generally could commit just fine even if they exceeded their original deadline. In other cases, you were of the opinion that one should never wait for a sync lease acquisition.
And keep in mind that we are supposed to refresh these leases async when they have less than a minute left... 
I'm not sure how Jordan is reproducing this in 25 seconds...

**Comment by spencerkimball** [comment]
Would be really nice if we could get a trace for the failing request.

**Comment by jordanlewis** [comment]
> I'm not sure how Jordan is reproducing this in 25 seconds...

That is a great question! It seems to indicate something weird is going on, doesn't it?

I will try to get a trace for one of these failing requests, but I don't know how easy it will be.

**Comment by petermattis** [comment]
Isn't it possible that @jordanlewis had previously run `tpcc` against his cluster which had grabbed various table leases, then waited a few minutes and started it up again such that one of the new txns tries to use the table lease just before it expired.

**Comment by tschottdorf** [comment]
The `drop` roachtest (which manually deletes all rows in a big table) is also seeing this.

**Comment by jordanlewis** [comment]
This is a blocker for reporting TPCC numbers. We need to get to the bottom of this and fix it. Saw another one after 9 minutes of an otherwise great looking, 3-second-max latency TPCC run.

**Comment by tschottdorf** [comment]
If I understand it correctly then the hypothesis is that we're giving a transaction a table lease that is relatively close to expired, and then the txn catches a retry with the new timestamp violating the deadline? (Even though other txns are probably already happily using another lease)? This looks like a bookkeeping problem, but a pretty bad one at that.
Are we sure that that's what's happening? When are leases actually renewed? Must be pretty close to the deadline if the relatively short-running TPCC txns get it. Is there maybe an obvious fix?

I agree with @jordanlewis that this needs to be addressed ASAP.

**Comment by vivekmenezes** [comment]
https://github.com/vivekmenezes/cockroach/commit/4ee05c3b2f3c3ef9983ba6c7a22a3f887cbc7d87

I'm not seeing this code panic which is good.

So this problem is either

1. table not touched in a long time and then used after 4-5 minutes (not seeing this on tpcc)
2. long running txn pushed by a minute
3. deadline code not being processed properly by kv layer

I'm not available for the next few hours and I'm cool with us putting in the retryable error fix so users do not see this. we can investigate this on the side and get to the bottom of this.

**Comment by petermattis** [comment]
The transaction deadline is only checked at commit time. I wonder if we could check the transaction deadline right before we commit the transaction. If the transaction is near the deadline (within N seconds), we could try to refresh the deadline by looping over all of the tables referenced by the transaction and checking to see if a new lease has been acquired for those tables and that table version has not changed.

**Comment by vivekmenezes** [comment]
that is something @andreimatei was proposing we do. I think it is doable that's one option (although it is not full proof). I think we still need the retry error for the rare situation that is not enough. We could also expose the RefreshTimestamp to sql and it could check before committing if the refresh timestamp > deadline

**Comment by vivekmenezes** [comment]
But I believe 1PC wouldn't give us the opportunity to refresh the deadline.

**Comment by petermattis** [comment]
> But I believe 1PC wouldn't give us the opportunity to refresh the deadline.

Why not? 1PC vs normal is immaterial for my proposal.

**Comment by benesch** [comment]
Ok, I finally tracked this down. The tl;dr is that I don't think we properly trigger lease refreshes when looking a table up by ID.

Here's what happens in TPC-C. In every transaction deadline error I've observed, the problem is caused by the `history` table's lease. The transaction's original timestamp will be before the history table's lease expires—in some cases, just 200ms before the expiration—but it'll encounter a serializability error when it tries to commit. The sender refreshes the timestamp, but the the timestamp is after the `history` lease expiration and the transaction correctly fails with a "deadline exceeded" error.

So why isn't the history table's lease getting refreshed? There's only one transaction in TPC-C, the payment transaction, that touches the history table directly. Roughly, that transaction executes these statements:

```sql
UPDATE warehouse...         -- 1
UPDATE district...          -- 2
SELECT ... FROM customer    -- 3
UPDATE customer...          -- 4
INSERT INTO history...      -- 5
```

Normally, query five would attempt to get a lease for the table with name `history`, triggering an async refresh of the `history` table's lease if there was less than a minute left before it expired. But query two accesses the `district` table, which has a foreign key to `history`. I don't quite understand why yet, but this forces a lease acquisition of the `history` table. This acquisition happens by ID, not name, and this code path does not check for whether the lease needs to be asynchronously refreshed. So when query five executes, the `history` table is already cached in the transaction's "table collection", and a lease refresh is never triggered.

The above timeline is derived from this trace of the payment transaction: https://gist.github.com/benesch/8e3df83ade0581333f8e7a9b19e4bb74

I also captured a nice clean failure of the new order transaction: https://gist.github.com/benesch/deaca885eb3007069e1b32e7275f2854

I'm preparing a solution now. I think triggering async refreshes of leases on the lookup-by-ID path, and not just the lookup-by-name path, will do the trick.

For anyone trying to decipher the traces, these are the additional log messages I added to my binary:

```
diff --git a/pkg/sql/lease.go b/pkg/sql/lease.go
index bb0e6d0b7..e1f8eb67d 100644
--- a/pkg/sql/lease.go
+++ b/pkg/sql/lease.go
@@ -857,6 +857,7 @@ func (t *tableState) acquireNodeLease(
 	if err != nil {
 		return nil, err
 	}
+	log.Warningf(ctx, "acquired a lease on table id=%d, expiration=%s", t.id, table.expiration)
 	t.mu.Lock()
 	defer t.mu.Unlock()
 	t.upsertLocked(ctx, table, m)
@@ -1246,18 +1247,24 @@ func nameMatchesTable(table *sqlbase.TableDescriptor, dbID sqlbase.ID, tableName
 func (m *LeaseManager) AcquireByName(
 	ctx context.Context, timestamp hlc.Timestamp, dbID sqlbase.ID, tableName string,
 ) (*sqlbase.TableDescriptor, hlc.Timestamp, error) {
+	log.Warningf(ctx, "acquiring lease on db=%d, table=%s", dbID, tableName)
+
 	// Check if we have cached an ID for this name.
 	tableVersion := m.tableNames.get(dbID, tableName, timestamp)
 	if tableVersion != nil {
+		log.Warningf(ctx, "lease found at ts %s, txn ts is %s", tableVersion.ModificationTime, timestamp)
 		if !timestamp.Less(tableVersion.ModificationTime) {
 
 			// Atomically check and begin a renewal if one has not already
 			// been set.
 
 			durationUntilExpiry := time.Duration(tableVersion.expiration.WallTime - timestamp.WallTime)
+			log.Warningf(ctx, "duration until lease expires is %s", durationUntilExpiry)
 			if durationUntilExpiry < m.LeaseStore.leaseRenewalTimeout {
-				if t := m.findTableState(tableVersion.ID, false /* create */); t != nil &&
-					atomic.CompareAndSwapInt32(&t.renewalInProgress, 0, 1) {
+				t := m.findTableState(tableVersion.ID, false /* create */)
+				log.Warningf(ctx, "table state is %+v", t)
+				if t != nil && atomic.CompareAndSwapInt32(&t.renewalInProgress, 0, 1) {
+					log.Warningf(ctx, "actually launching an async renewal")
 					// Start the renewal. When it finishes, it will reset t.renewalInProgress.
 					if err := t.stopper.RunAsyncTask(context.Background(),
 						"lease renewal", func(ctx context.Context) {
@@ -1273,6 +1280,7 @@ func (m *LeaseManager) AcquireByName(
 
 			return &tableVersion.TableDescriptor, tableVersion.expiration, nil
 		}
+		log.Warningf(ctx, "no lease for this timestamp; acquiring a new lease")
 		if err := m.Release(&tableVersion.TableDescriptor); err != nil {
 			return nil, hlc.Timestamp{}, err
 		}
@@ -1350,6 +1358,7 @@ func (m *LeaseManager) AcquireByName(
 			return nil, hlc.Timestamp{}, sqlbase.ErrDescriptorNotFound
 		}
 	}
+	log.Warningf(ctx, "no cached descriptor found; hit the kv db successfully")
 	return table, expiration, nil
 }
 
diff --git a/pkg/sql/table.go b/pkg/sql/table.go
index 3aa7fc2f0..d98d458d9 100644
--- a/pkg/sql/table.go
+++ b/pkg/sql/table.go
@@ -301,7 +301,7 @@ func (tc *TableCollection) getTableVersion(
 	}
 	tc.timestamp = origTimestamp
 	tc.leasedTables = append(tc.leasedTables, table)
-	log.VEventf(ctx, 2, "added table '%s' to table collection", tn)
+	log.VEventf(ctx, 2, "added table '%s' to table collection with deadline %s", tn, expiration)
 
 	// If the table we just acquired expires before the txn's deadline, reduce
 	// the deadline. We use OrigTimestamp() that doesn't return the commit timestamp,
@@ -366,7 +366,7 @@ func (tc *TableCollection) getTableVersionByID(
 	}
 	tc.timestamp = origTimestamp
 	tc.leasedTables = append(tc.leasedTables, table)
-	log.VEventf(ctx, 2, "added table '%s' to table collection", table.Name)
+	log.VEventf(ctx, 2, "added table '%s' to table collection with deadline %s", table.Name, expiration)
 
 	// If the table we just acquired expires before the txn's deadline, reduce
 	// the deadline. We use OrigTimestamp() that doesn't return the commit timestamp,

```

**Comment by vivekmenezes** [comment]
Nice cache nikhil. Yeah the descriptor lookup by Id happens for fks since
those lookups happen through other descriptors. Great!

**Comment by spencerkimball** [comment]
Wow, was that an intentional pun, @vivekmenezes? Bravo, regardless.

**Comment by petermattis** [comment]
Does this also explain the failures seen in the drop roachtest? We're deleting from tables that have foreign keys, so perhaps.

Cc @tschottdorf

**Comment by benesch** [comment]
You'd still need another read to come along and push the DELETE FROM
transaction. I don't see what would cause that in the test itself, but
perhaps there's some automatic system txns that come along?

**Comment by jordanlewis** [comment]
I hate to do this, but here's 37 minutes into a 3-node run with 1250 warehouses:

```
_time______opName__ops/s(inst)__ops/s(cum)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
37m1s    newOrder        172.6       249.2   2684.4   7784.6  12884.9  36507.2  40802.2
37m2s    newOrder        233.6       249.2   2818.6   7784.6  13421.8  25769.8  38654.7
37m3s    newOrder        161.7       249.2   2550.1   7516.2  12348.0  23622.3  36507.2
37m4s    newOrder        221.0       249.2   2281.7   5905.6   8053.1  21474.8  24696.1
37m5s    newOrder        217.7       249.1   2550.1   7784.6  13421.8  32212.3  42949.7
37m6s    newOrder        215.0       249.1   2415.9   6979.3  10737.4  30064.8  45097.2
2018/03/19 02:03:54 error in payment: pq: TransactionStatusError: transaction deadline exceeded
```

Is it possible that this error can still be generated under normal circumstances due to very slow transactions? Is that okay? Should this error still be marked retryable, despite the bugfix?

**Comment by vivekmenezes** [comment]
After this fix there are three circumstances under which this can still happen

1. cached table descriptor that isn't read for a while and happens to get read close to the deadline (I'll send out a fix for this case)
2. transactions started just before a schema change as described in these tests https://github.com/cockroachdb/cockroach/pull/23681 
3. long running transactions that get pushed past the deadline.

I feel that we should merge spencer's change that makes this a retryable error so that case 2 is handled properly.  I suppose case 3 is worrisome because making this a retryable error could result in an infinite retry loop for a long transaction. Are they safeguards to prevent infinite retries of transactions?

**Comment by benesch** [comment]
Oops, didn't mean for my PR to close this issue. Agreed we should make these errors retryable, though I'm in Andrei's camp: SQL should be marking these transactions as retryable, not KV. I'm happy to defer that opinion until 2.1, though.

> cached table descriptor that isn't read for a while and happens to get read close to the deadline (I'll send out a fix for this case)

@vivekmenezes, I'm curious what fix you have in mind. The only solution I can come up with is reintroducing a MinSafeDuration.

**Comment by bdarnell** [comment]
> I suppose case 3 is worrisome because making this a retryable error could result in an infinite retry loop for a long transaction. Are they safeguards to prevent infinite retries of transactions?

No, and that's why I am opposed to making this a retryable error. We need to either be able to extend the transaction's deadline while it is running or ensure that there is enough time left on the lease when the transaction starts (synchronously renewing it if necessary).

**Comment by bdarnell** [comment]
I would also be in favor of extending the lease duration constants - enabling live transactions to complete is much more important than reducing the delay before a schema change can run after an unclean node shutdown.

**Comment by andreimatei** [comment]
> 1. cached table descriptor that isn't read for a while and happens to get read close to the deadline (I'll send out a fix for this case)

What fix, exactly?

**Comment by vivekmenezes** [comment]
I'm at an off-site but I have a change that essentially changes the
implementation of hasexpired()

**Comment by benesch** [comment]
I'm punting this out of 2.0 because I think we've done all we can for now.

**Comment by vivekmenezes** [comment]
Here's the plan to put a nail in the coffin on this issue by completely eliminating the need for deadlines. @andreimatei to review.

1. Disallow cached descriptors that are not leased. We introduced the ability to cache descriptors from arbitrary versions back in time but this is turning out increase the cognitive load on the system with very little benefit. The original thinking is that it would allow AS OF SYSTEM queries to use cached descriptors, but that is really not a use case worth supporting. If in the future we think we want to support this cache it should come from some real user need. It's not hard to support an AS OF SYSTEM TIME cache that is managed separately. Doing this ensures that transactions use table leases and with 2 below we will not need to impose a deadline for fear of a descriptor changing beyond a timestamp. This will also fix https://github.com/cockroachdb/cockroach/issues/19925 and eliminate all the deadline errors seen in #23681

2. move to epoch based descriptor leases to allow long running transactions holding leases to not have to impose a deadline (https://github.com/cockroachdb/cockroach/issues/23978)

3. remove notion of deadline from kv code!

**Comment by andreimatei** [comment]
> Doing this ensures that transactions use table leases and with 2 below we will not need to impose a deadline for fear of a descriptor changing beyond a timestamp. 

Can you explain this a bit more? In the face of schema changes, there need to be a deadlines for txns... No?

**Comment by vivekmenezes** [comment]
I think I'd like to add the originating node and the liveness epoch of the node to a transaction, so today where we check for deadlines it checks if the originating node is still alive for that epoch (I'm assuming this can be looked up via a cache)

The lease will be associated with a node (as is done today) and a liveness epoch for the node (same as txn using the lease).

The schema change can move forward with incrementing the version to V+2, when there are no valid leases for versions V and below. A lease is invalid when it is associated with a node and an epoch that is past or the lack of liveness for a long time (this needs to be ironed out, I've not familiar with replica leases and I need to understand that better)

A node will give up a lease on version V when it is no longer in use and when it has a lease on V + 1. It will also periodically (because gossip is unreliable) read the descriptor to check if the version has been incremented and thus needs to acquire a new lease.

If this is all looking encouraging, I can move forward with a prototype and update to the RFC to iron this out.

**Comment by petermattis** [comment]
> A node will give up a lease on version V when it is no longer in use and when it has a lease on V + 1. It will also periodically (because gossip is unreliable) read the descriptor to check if the version has been incremented and thus needs to acquire a new lease.

If I'm understanding your proposal correctly, a long running transaction can prevent a scheme change from running for an arbitrarily long period of time. That is a significant departure from the current behavior. Perhaps worth considering, but it seems that should be the focus of the proposal, not the mechanism by which to achieve that effect.

**Comment by vivekmenezes** [comment]
@petermattis I didn't quite follow your comment about 1PC that you had made a few comments above. I thought 1PC is handed over as a single batch that includes the commit message and the commit timestamp is unknown to the coordinator node, or known after the commit has actually happened. So if the commit timestamp passes the deadline, the transaction will fail.

**Comment by tschottdorf** [comment]
What's the status? @HeikoOnnebrink was curious where to go from here.

**Comment by andreimatei** [comment]
Don't have this paged in any more, but my impression was that this should be quite rare now - or at least relegated to really long transactions. Is anyone seeing these errors still?

**Comment by vivekmenezes** [comment]
@tschottdorf we plan to fix this for 2.1. It's not in the current monthly milestone. It should be fixable!

**Comment by tim-o** [comment]
@andreimatei, @HeikoOnnebrink ran into this while testing TPC-C on a 16 node cluster. AFAIK, it's rare, but it is happening when trying to replicate our benchmark results.

**Comment by nvanbenschoten** [comment]
We have been seeing this more frequently in tests (#26529, #27688, #29258) and I also just saw it on a long-running TPC-C cluster. I think the issue is related to some of the refresh span optimizations added in #21140 and #22477. Specifically, tracking refresh spans and the optimizations that permits allows us to break the following invariant:

https://github.com/cockroachdb/cockroach/blob/d703a5775dfd0d7b0af36a480d117d870c3d60dc/pkg/roachpb/api.proto#L488-L493

I can see two ways that this would lead to us hitting these errors:
1. we don't consult the transaction deadline when calling [`CanTransactionRetryAtRefreshedTimestamp`](https://github.com/cockroachdb/cockroach/blob/d703a5775dfd0d7b0af36a480d117d870c3d60dc/pkg/roachpb/data.go#L1167) and refreshing a transaction's timestamp, meaning that the `txnSpanRefresher` could move a transaction's `Timestamp` and `RefreshTimestamp` past its deadline and the transaction would not trigger a retry error when attempting to commit, instead getting hit with the more serious `transaction deadline exceeded` error.
2. we don't consult the transaction deadline when committing a transaction and checking whether we  [`canForwardSerializableTimestamp`](https://github.com/cockroachdb/cockroach/blob/d703a5775dfd0d7b0af36a480d117d870c3d60dc/pkg/storage/batcheval/cmd_end_transaction.go#L396). This again allows the commit to avoid triggering a retry error, but instead leaves us exposed to a `transaction deadline exceeded` error which we're [not supposed to be able to hit](https://github.com/cockroachdb/cockroach/blob/d703a5775dfd0d7b0af36a480d117d870c3d60dc/pkg/storage/batcheval/cmd_end_transaction.go#L278).


In both cases, we're trading what would have been a client-side retryable error before #21140 and #22477 for a non-retryable transaction status error after. I see two alternative ways of fixing this:
- we prevent the forwarding of txn timestamps past the txn deadline when attempting to avoid retryable errors, accepting that the retryable errors can't be avoided. This might be tricky because we'll need to have access to the deadline wherever we may want to push a transaction's timestamp, such as in the `txnSpanRefresher`.
- we can make `transaction deadline exceeded` errors retryable themselves. I'm actually not sure why we don't already do this.

@spencerkimball @andreimatei I'd like to get your input on this. Am I missing anything?

**Comment by vivekmenezes** [comment]
@nvanbenschoten I believe we've not made transaction deadline a retryable error because in the past we've felt that doing so would paper over some hidden bug. I feel with 2.2 plans for epoch based leases: https://github.com/cockroachdb/cockroach/issues/23978 and delayed assignment of a deadline:  https://github.com/cockroachdb/cockroach/issues/24041 the situation will get better but will not completely fixed this issue: schema changes can come along forcing short deadlines on the use of older versions of cached descriptors.

But, given a recent interim solution to periodically refresh a set number of tables through: https://github.com/cockroachdb/cockroach/pull/28725 and also given the complete lack of schema changes in the TPCC tests it is indeed surprising we are seeing it in this scenario.  I worry that we have a hidden problem somewhere. Perhaps at a minimum we can add some logging to log the origtimestamp and the deadline whenever this error is seen by the server so that we can see that they are close to each other or not?

**Comment by andreimatei** [comment]
The only thing I have to say on the topic is that, ideally, one way or another, we'd make it so that, in the absence of schema changes, a deadline would never be hit (or at least not hit unless the commit batch itself takes a long time to evaluate). Either being smarter about how the deadline is assigned (i.e. if a lease was refreshed while a txn using that descriptor was running, the lease expiration time from before the refresh should not affect the deadline), or moving to a scheme where txns are tied to a liveness epoch and leases do not otherwise have their own expiration time (like Vivek has been floating recently) should achieve this.

In the meantime, while we have these deadlines, as Nathan says, pushing the txn above deadline makes no sense and all the pusher call sites should avoid doing so (except I think the deadline is technically an attribute of a commit request and not of a txn, so some callsites might have a difficulty?).

Finally, given the specifics of TPCC (and particularly now that #28725 has sneaked in and so now the first 50 tables or something have their leases constantly refreshed), it would seem that the only way to hit the deadline in TPCC is if a txn has been pushed by something on the order of half a leases' duration or something - which is minutes - which I guess sounds a bit suspicious?

**Comment by bdarnell** [comment]
> we can make transaction deadline exceeded errors retryable themselves. I'm actually not sure why we don't already do this.

Transaction deadlines should always be fairly distant in the future. If the deadline is hit, the transaction has already run for a long time. Without epoch/refreshing leases, the restarted transaction would presumably have to do the same amount of work again and most likely hit the deadline again (which will be the same distance in the future). 

With epoch-based leases (which would mean that we should only experience deadlines due to schema changes), hitting the deadline should be retryable because after the schema change has passed we'll have more time on the retry. 

In the meantime, we'd ideally check the deadline when pushing the transaction, turning this into a retryable error at that level.

**Comment by bdarnell** [comment]
> Finally, given the specifics of TPCC (and particularly now that #28725 has sneaked in and so now the first 50 tables or something have their leases constantly refreshed), it would seem that the only way to hit the deadline in TPCC is if a txn has been pushed by something on the order of half a leases' duration or something - which is minutes - which I guess sounds a bit suspicious?

And yes, the fact that this is becoming more common in the absence of schema changes suggests that something else is going wrong and we should look into that.

**Comment by nvanbenschoten** [comment]
One theory I have for why this is becoming common for these bulk-loading txns is that they're running into the closed timestamp tracker's new minimum timestamp. This minimum timestamp enforces an upper bound on how far in the past a transaction can propose a write. The default value for this window is 5s (see `kv.closed_timestamp.target_duration`). It's possible that these batch writes when bulk loading are so large that they never complete in 5 seconds, so they keep getting pushed indefinitely. I'm performing a TPC-C 1000 load now, so I'll watch for this issue again. If I see it, I'll try bumping that setting and seeing if it helps.

**Comment by nvanbenschoten** [comment]
I think we'd only expect the proposal timestamp to be more than 5 seconds in the past if one of these implicit txns was split up by the DistSender and not all of it was sent in parallel because of the concurrency limit. I'm running a binary with this limit bumped by an order of magnitude, so it's possible I won't see this issue materialize. I'll run this side-by-side with a normal beta binary and see if there are any differences.

**Comment by vivekmenezes** [comment]
@nvanbenschoten how about reducing it and seeing if this problem is indeed more reproducible with the value set to say 1s

**Comment by awoods187** [comment]
Ran into this on a 4 node tpc-c 2k on AWS `andy-tpcc-8`
```
Error: error in newOrder: ERROR: TransactionStatusError: transaction deadline exceeded (REASON_UNKNOWN) (SQLSTATE XX000)
Error:  exit status 1
```

**Comment by nvanbenschoten** [comment]
@andreimatei @vivekmenezes Now that we've determined the root cause of these long transactions, do we feel comfortable making transaction deadline exceeded errors retriable? The fix for the long transactions isn't trivial and I don't think it should be a blocker for avoiding these errors.

**Comment by andreimatei** [comment]
Well I think the first thing to do here is to ensure that we never push a txn above its deadline.
Then, yeah, I think we should make the deadline errors retriable.

**Comment by vivekmenezes** [comment]
@andreimatei are you saying return a retryable whenever we detect we're
pushing a transaction beyond it's deadline?

**Comment by andreimatei** [comment]
> @andreimatei are you saying return a retryable whenever we detect we're
pushing a transaction beyond it's deadline?

Yes. The question, of course, is who should create the errors, since "pushing" is a bit of a conflated term. For one, I think server-side code should probably stay unchanged, and all the handling of deadline-related issues should be relegated to the client.
For refreshes, I think `CanTransactionRetryAtRefreshedTimestamp` should probably return `false` here:
https://github.com/cockroachdb/cockroach/blob/36eb40c7a5d6c10f72c694c56612d4aa3ae78099/pkg/kv/txn_interceptor_span_refresher.go#L197

Then, I think `IsSerializablePushAndRefreshNotPossible()` should also become aware of the deadline.
https://github.com/cockroachdb/cockroach/blob/e4a2a4c29c4b0ed164c08684545aeb26a5e8759b/pkg/kv/txn_coord_sender.go#L1062

Then I think code around here should also always detect txns pushed above the deadline and return an error eagerly, although, as with all eager versus deferred error discussions, this is a judgement call.
https://github.com/cockroachdb/cockroach/blob/36eb40c7a5d6c10f72c694c56612d4aa3ae78099/pkg/kv/txn_interceptor_span_refresher.go#L151

**Comment by vivekmenezes** [comment]
I'll plan on giving this a try in our next monthly milestone.

**Comment by vivekmenezes** [comment]
One plan here is to always set the deadline at commit time so that the transaction uses the deadline of a later refreshed lease with an updated (refreshed) deadline that is more likely to succeed. This can be achieved by ensuring that the transaction Commit API always uses a deadline instead of making the deadline a property of the transaction itself. At commit time the sql layer can call the API with the latest refreshed deadline for all the descriptors used in the transaction.

The above will also get rid of needing to pass the transaction as a flag to the schema resolver API.

**Comment by danhhz** [comment]
FWIW, I just hit this in a local run of the `tpcc/nodes=3/w=max` roachtest running a binary at sha 72edf20311c08d8347f5e0055b1612a68826c894.

**Comment by vivekmenezes** [comment]
@danhhz any chance you can add the error message because it now provides more information. Thanks

**Comment by danhhz** [comment]
Yup. I actually saw this twice. Here are both:
```
error in payment: ERROR: TransactionStatusError: transaction deadline exceeded by 1.653254848s (1550089726.208602717,1 > 1550089724.555347869,0), original timestamp 1m15.162854897s ago (1550089651.045747820,0): id=cf75031c key=/Table/56/1/428/0 rw=true pri=0.01464384 stat=PENDING epo=0 ts=1550089726.208602717,1 orig=1550089651.045747820,0 max=1550089651.045747820,0 wto=false seq=7 (REASON_UNKNOWN) (SQLSTATE XX000)
```
```
error in payment: ERROR: TransactionStatusError: transaction deadline exceeded by 7.360572099s (1550089138.588048773,1 > 1550089131.227476674,0), original timestamp 2m23.181141999s ago (1550088995.406906774,0): id=26d9e30d key=/Table/61/1/414/0 rw=true pri=0.07253801 stat=PENDING epo=0 ts=1550089138.588048773,1 orig=1550088995.406906774,0 max=1550088995.407374179,0 wto=false seq=7 (REASON_UNKNOWN) (SQLSTATE XX000)
```

**Comment by vivekmenezes** [comment]
thanks. This surely indicates to me that I need to first extend the deadline (reducing the chance of this error) before making this a retryable error.

**Comment by vivekmenezes** [comment]
@andreimatei I've confirmed that the tpcc tests in the presence of schema changes where schema versions are changes will retry transactions when they see this error. I think we should close this issue.

**Comment by andreimatei** [comment]
Well there's a lot of history in here about how we shouldn't be pushing transactions over their deadline
https://github.com/cockroachdb/cockroach/issues/18684#issuecomment-434729550

and also an analysis of very long transactions: https://github.com/cockroachdb/cockroach/issues/18684#issuecomment-420477586

If we close this, we should be opening separate issues for those I think.

**Comment by jordanlewis** [comment]
I don't think this issue is closeable until random workloads don't see "transaction deadline exceeded" anymore. https://github.com/cockroachdb/cockroach/issues/35337#issuecomment-473387718

**Comment by vivekmenezes** [comment]
@andreimatei I've created an issue for early detection of the deadline exceeded condition. The other long running transaction analysis is no longer applicable.

@jordanlewis we are no longer seeing test flakes

Closing this issue!

**Comment by jordanlewis** [comment]
Could it be true? The troll is slain? Congratulations!

**Comment by andreimatei** [comment]
@nvanbenschoten do you want to extract the intent resolving pathology in https://github.com/cockroachdb/cockroach/issues/18684#issuecomment-428433335 into a separate issue?

**Comment by nvanbenschoten** [comment]
Pulled into https://github.com/cockroachdb/cockroach/issues/36876.
